pkg/volume/storageos/storageos_util.go

```diff
@@ -0,0 +1,398 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageos
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/volume"
+
+	"github.com/golang/glog"
+	storageosapi "github.com/storageos/go-api"
+	storageostypes "github.com/storageos/go-api/types"
+)
+
+const (
+	losetupPath   = "losetup"
+	apiSecretName = "storageos-api"
+
+	modeBlock deviceType = iota
+	modeFile
+	modeUnsupported
+
+	ErrDeviceNotFound     = "device not found"
+	ErrDeviceNotSupported = "device not supported"
+	ErrNotAvailable       = "not available"
+)
+
+type deviceType int
+
+// storageosVolume describes a provisioned volume
+type storageosVolume struct {
+	ID          string
+	Name        string
+	Namespace   string
+	Description string
+	Pool        string
+	Size        int
+	Labels      map[string]string
+	FSType      string
+}
+
+type storageosAPIConfig struct {
+	apiAddr    string
+	apiUser    string
+	apiPass    string
+	apiVersion string
+}
+
+type apiConfigGetter interface {
+	GetAPIConfig() *storageosAPIConfig
+}
+
+type namespacedSecret struct {
+	namespace  string
+	secretName string
+	host       volume.VolumeHost
+}
+
+// GetAPIConfig checks the namespace for the api config, then default namespace.
+// Returns default config if no secrets found.
+func (a *namespacedSecret) GetAPIConfig() *storageosAPIConfig {
+	c := &storageosAPIConfig{
+		apiAddr:    defaultAPIAddress,
+		apiUser:    defaultAPIUser,
+		apiPass:    defaultAPIPassword,
+		apiVersion: defaultAPIVersion,
+	}
+
+	keys, err := a.host.GetKubeClient().Core().Secrets(a.namespace).Get(apiSecretName, metav1.GetOptions{})
```

Comment:
To add to this. When fetching a secret for a volume, the namespace used should always be the same as the pod and not user configurable -- this is to prevent a user in one namespace from being able to steal a secret from another namespace. Because of this you may want to decouple your StorageOS namespaces from k8s namespaces.